src/cockpit/polkit.py

```diff
@@ -100,56 +100,60 @@ async def _communicate(self, process: asyncio.subprocess.Process):
                 break
 
 
-async def register_agent(responder):
-    """Register polkit agent.
+class RegisterAgent:
+    """Register polkit agent when required
 
-    Returns an opaque object which needs to be kept during the lifetime of the agent.
-    After that, pass it to @unregister_agent().
-    """
-    try:
-        system_bus = bus.Bus.default_system()
-    except OSError as e:
-        logger.debug('cannot connect to system bus, not registering polkit agent: %s', e)
-        return None
-
-    try:
-        # may refine that with a D-Bus call to logind
-        subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
-    except KeyError:
-        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
-        return None
-
-    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
-    agent_slot = system_bus.add_object('/', agent_object)
-
-    # register agent
-    locale_name = locale.setlocale(locale.LC_MESSAGES, '')
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'RegisterAuthenticationAgent',
-        '(sa{sv})ss',
-        subject, locale_name, '/')
-    logger.debug('Registered agent for %r and locale %s', subject, locale_name)
-    return {'subject': subject, 'slot': agent_slot}
-
-
-async def unregister_agent(agent):
-    """Unregister polkit agent.
-
-    You have to do this explicitly (not just rely on garbage collection), otherwise
-    the next register_agent() will fail with "there is already a registered agent".
-
-    This receives the agent object returned by @register_agent().
+    This checks #argv if it contains "pkexec"; if so, it registeres a polkit agent,
+    otherwise does nothing.
+
+    Use this as a context manager to ensure that the agent gets unregistered again.
     """
-    del agent['slot']
-    system_bus = bus.Bus.default_system()
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'UnregisterAuthenticationAgent',
-        '(sa{sv})s',
-        agent['subject'], '/')
-    logger.debug('Unregistered agent for %r', agent['subject'])
+    def __init__(self, responder: InteractionResponder, argv: Sequence[str]):
+        self.responder = responder
+        self.agent_slot = None
+        self.needs_agent = 'pkexec' in argv
```

Comment:
I don't like this very much :/

I think I'd like to try to polyfill `AsyncExitStack`.  Something like:

```python
  class AsyncExitStack:   
      async def __aenter__(self):
          self.cms = []
          return self
  
      async def enter_async_context(self, cm):
          result = await cm.__aenter__()
          self.cms.add(cm)
          return result
          
      async def __aexit__(self, *exc_details):
          for cm in self.cms:
              cm.__aexit__() 
```

would get us enough functionality to be able to write it like:

```python
      async def do_connect_transport(self):
          async with AsyncExitStack() as stack:
              if 'pkexec' in self.argv:
                  await stack.enter_async_context(polkit.Agent(self.responder))
              ...

```

Reply:
This looks nicer indeed, thanks! I implemented that approach.